File: backend/app/models/membership.py
"""membership_tiers — tier configuration (M1).

Config rows for the three tiers (Special/Founding/Standard) so an admin can manage
their parameters. `code` matches the `MembershipTier` enum stored on `users`.
`revenue_share_bps` is the reviewer's revenue share in basis points (e.g. 3000 = 30%)
and feeds the tier-based revenue split (M2).
"""

from __future__ import annotations

from sqlalchemy import Boolean, Enum, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, Timestamps, UUIDPrimaryKey
from app.models.enums import MembershipTier


class MembershipTierConfig(Base, UUIDPrimaryKey, Timestamps):
    __tablename__ = "membership_tiers"

    code: Mapped[MembershipTier] = mapped_column(
        Enum(MembershipTier, name="membership_tier"), unique=True, nullable=False
    )
    name: Mapped[str] = mapped_column(String(80), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    # Reviewer revenue share in basis points (0..10000).
    revenue_share_bps: Mapped[int] = mapped_column(Integer, nullable=False, server_default="3000")
    # Lower number = paid earlier in the payout schedule (M3).
    payout_priority: Mapped[int] = mapped_column(Integer, nullable=False, server_default="100")
    benefits: Mapped[dict | None] = mapped_column(JSONB)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, server_default="true")

File: lib/membership.ts
import "server-only";

import { apiFetch } from "./api/client";

/**
 * Membership tiers (ADR-012). These are *assigned status levels*, not purchased
 * plans — there is no checkout anywhere in this flow. A tier controls two things:
 * `revenue_share_bps`, the reviewer's cut of affiliate revenue in basis points,
 * and `payout_priority`, the order payouts are scheduled in.
 *
 * Tier config is DB-managed and moderator-editable, so this page reads it live
 * rather than hardcoding the numbers — the M2 revenue split and M3 payout
 * scheduler read the same rows.
 */

export type Tier = {
  id: string;
  code: "special" | "founding" | "standard";
  name: string;
  description: string | null;
  revenue_share_bps: number;
  payout_priority: number;
  benefits: Record<string, unknown> | null;
  is_active: boolean;
};

/**
 * The public tier table, ordered most-generous first so the page reads as a
 * ladder. Returns [] if the backend is unreachable — the page renders its
 * explanatory copy either way rather than erroring out.
 */
export async function getTiers(): Promise<Tier[]> {
  try {
    const tiers = await apiFetch<Tier[]>("/api/v1/membership-tiers");
    return tiers
      .filter((t) => t.is_active)
      .sort((a, b) => b.revenue_share_bps - a.revenue_share_bps);
  } catch {
    return [];
  }
}

/** 7500 bps → "75%". Basis points are integers; format, never round for math. */
export function bpsToPercent(bps: number): string {
  const pct = bps / 100;
  return `${Number.isInteger(pct) ? pct : pct.toFixed(2)}%`;
}

/** Human label for the payout-priority integer (lower runs earlier). */
export function priorityLabel(priority: number): string {
  if (priority <= 0) return "First in line";
  if (priority === 1) return "Early";
  return "Standard queue";
}

/**
 * The benefits JSONB is admin-editable free-form, so it can hold anything.
 * Render only string/number/boolean leaves as "Key: value" bullets and drop
 * nested structures rather than dumping JSON at the reader.
 */
export function benefitLines(benefits: Record<string, unknown> | null): string[] {
  if (!benefits) return [];
  const lines: string[] = [];
  for (const [key, value] of Object.entries(benefits)) {
    const label = key.replace(/_/g, " ").replace(/^\w/, (c) => c.toUpperCase());
    if (typeof value === "boolean") {
      if (value) lines.push(label);
    } else if (typeof value === "string" || typeof value === "number") {
      lines.push(`${label}: ${value}`);
    }
  }
  return lines;
}
